File: tests/Snapik.App.Imaging.Tests/SoundVolumeCurveTests.cs
using Snapik.App;

namespace Snapik.App.Imaging.Tests;

// The slider of the settings is squared before it reaches the player, so that its travel is spent
// where the ear notices. The three gains are the ones the three sounds are mixed with.
public sealed class SoundVolumeCurveTests
{
    private const double Shutter = 0.6;
    private const double Tick = 0.25;
    private const double Copied = 0.7;

    [Fact]
    public void The_top_of_the_slider_stays_where_it_was() =>
        Assert.Equal(Shutter, SoundVolumeCurve.Amplitude(100, Shutter), 6);

    [Fact]
    public void Half_the_slider_is_a_quarter_of_the_sound() =>
        Assert.Equal(0.15, SoundVolumeCurve.Amplitude(50, Shutter), 6);

    [Fact]
    public void A_quarter_of_the_slider_is_a_sixteenth_of_the_sound() =>
        Assert.Equal(0.0375, SoundVolumeCurve.Amplitude(25, Shutter), 6);

    [Fact]
    public void Zero_is_silence_and_not_a_whisper()
    {
        Assert.Equal(0, SoundVolumeCurve.Amplitude(0, Shutter), 6);
        Assert.Equal(0, SoundVolumeCurve.Amplitude(0, Copied), 6);
    }

    [Fact]
    public void A_value_outside_the_slider_is_pulled_back_onto_it()
    {
        Assert.Equal(0, SoundVolumeCurve.Amplitude(-5, Shutter), 6);
        Assert.Equal(Shutter, SoundVolumeCurve.Amplitude(250, Shutter), 6);
    }

    [Fact]
    public void The_curve_only_ever_rises()
    {
        foreach (var gain in new[] { Tick, Shutter, Copied })
            for (var volume = 1; volume <= 100; volume++)
                Assert.True(SoundVolumeCurve.Amplitude(volume, gain) > SoundVolumeCurve.Amplitude(volume - 1, gain));
    }
}

File: macos/Tests/SnapikCoreTests/SoundVolumeCurveTests.swift
// Port of `tests/Snapik.App.Imaging.Tests/SoundVolumeCurveTests.cs`, SPEC-DELTA-5 §3.3.
//
// The three gains are the ones the interface carries: the shutter 0.6, the tick of the strip 0.25
// and the answer of a copy 0.7 (`UiSoundService`).
import XCTest

@testable import SnapikCore

final class SoundVolumeCurveTests: XCTestCase {
    private static let gains: [Double] = [0.6, 0.25, 0.7]

    func test_The_top_of_the_slider_is_the_gain_of_the_sound_itself() {
        XCTAssertEqual(0.6, SoundVolumeCurve.amplitude(volume: 100, gain: 0.6), accuracy: 1e-6)
    }

    /// Half the travel is a quarter of the amplitude, a quarter of it is a sixteenth: that is the
    /// whole point of the square.
    func test_Half_the_travel_is_a_quarter_of_the_amplitude() {
        XCTAssertEqual(0.15, SoundVolumeCurve.amplitude(volume: 50, gain: 0.6), accuracy: 1e-6)
        XCTAssertEqual(0.0375, SoundVolumeCurve.amplitude(volume: 25, gain: 0.6), accuracy: 1e-6)
    }

    func test_Nothing_at_the_bottom_is_silence_whatever_the_gain() {
        XCTAssertEqual(0, SoundVolumeCurve.amplitude(volume: 0, gain: 0.6), accuracy: 1e-6)
        XCTAssertEqual(0, SoundVolumeCurve.amplitude(volume: 0, gain: 0.25), accuracy: 1e-6)
    }

    /// A number out of the range of the slider is a number a hand-edited file holds: it is clamped,
    /// not believed.
    func test_A_volume_outside_the_slider_is_clamped_to_its_ends() {
        XCTAssertEqual(0, SoundVolumeCurve.amplitude(volume: -5, gain: 0.6), accuracy: 1e-6)
        XCTAssertEqual(0.6, SoundVolumeCurve.amplitude(volume: 250, gain: 0.6), accuracy: 1e-6)
    }

    /// Every step of the slider is louder than the one before it, on all three gains: a curve that
    /// dipped anywhere would make the slider feel broken even where the numbers are right.
    func test_The_curve_only_rises() {
        for gain in Self.gains {
            var previous = SoundVolumeCurve.amplitude(volume: 0, gain: gain)
            for volume in 1...100 {
                let current = SoundVolumeCurve.amplitude(volume: volume, gain: gain)
                XCTAssertGreaterThan(current, previous, "\(gain) \(volume)")
                previous = current
            }
            XCTAssertEqual(gain, previous, accuracy: 1e-6)
        }
    }
}
